// react-stick@5.0.3/lib/utils/fit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fitsOnBottom = exports.fitsOnTop = exports.fitsOnLeft = exports.fitsOnRight = exports.isPositionedToLeft = exports.isPositionedToRight = exports.isPositionedToBottom = exports.isPositionedToTop = void 0;
const isPositionedToTop = (position) => {
    const [positionMarker] = position.split(' ');
    return positionMarker === 'top';
};
exports.isPositionedToTop = isPositionedToTop;
const isPositionedToBottom = (position) => {
    const [positionMarker] = position.split(' ');
    return positionMarker === 'bottom';
};
exports.isPositionedToBottom = isPositionedToBottom;
const isPositionedToRight = (position) => {
    const positionMarker = position.split(' ')[1];
    return positionMarker === 'right';
};
exports.isPositionedToRight = isPositionedToRight;
const isPositionedToLeft = (position) => {
    const positionMarker = position.split(' ')[1];
    return positionMarker === 'left';
};
exports.isPositionedToLeft = isPositionedToLeft;
const fitsOnRight = (nodeRef, anchorRef) => {
    const { width: nodeWidth } = nodeRef.getBoundingClientRect();
    const { right: anchorRight } = anchorRef.getBoundingClientRect();
    return anchorRight + nodeWidth <= window.innerWidth;
};
exports.fitsOnRight = fitsOnRight;
const fitsOnLeft = (nodeRef, anchorRef) => {
    const { width: nodeWidth } = nodeRef.getBoundingClientRect();
    const { left: anchorLeft } = anchorRef.getBoundingClientRect();
    return anchorLeft - nodeWidth >= 0;
};
exports.fitsOnLeft = fitsOnLeft;
const fitsOnTop = (nodeRef, anchorRef) => {
    const { height: nodeHeight } = nodeRef.getBoundingClientRect();
    const { top: anchorTop } = anchorRef.getBoundingClientRect();
    return anchorTop - nodeHeight >= 0;
};
exports.fitsOnTop = fitsOnTop;
const fitsOnBottom = (nodeRef, anchorRef) => {
    const { height: nodeHeight } = nodeRef.getBoundingClientRect();
    const { bottom: anchorBottom } = anchorRef.getBoundingClientRect();
    return anchorBottom + nodeHeight <= window.innerHeight;
};
exports.fitsOnBottom = fitsOnBottom;
