// .ignored_nextra/dist/chunk-L2E34PAX.mjs
import {
  LOCALE_REGEX
} from "./chunk-P7XAM56D.mjs";

// src/utils.ts
import path from "path";
import title from "title";
function parseFileName(filePath) {
  var _a;
  const { name, ext } = path.parse(filePath);
  const locale = ((_a = name.match(LOCALE_REGEX)) == null ? void 0 : _a[1]) || "";
  return {
    name: locale ? name.replace(LOCALE_REGEX, "") : name,
    locale,
    ext
  };
}
var parseJsonFile = (content, path2) => {
  try {
    return JSON.parse(content);
  } catch (err) {
    console.error(
      `[nextra] Error parsing ${path2}, make sure it's a valid JSON`,
      err
    );
    return {};
  }
};
function truthy(value) {
  return !!value;
}
function normalizeMeta(meta) {
  return typeof meta === "string" ? { title: meta } : meta;
}
function sortPages(pages, locale) {
  return pages.filter((item) => item.kind === "Folder" || item.locale === locale).map((item) => {
    var _a, _b;
    return {
      name: item.name,
      date: "frontMatter" in item && ((_a = item.frontMatter) == null ? void 0 : _a.date),
      title: "frontMatter" in item && ((_b = item.frontMatter) == null ? void 0 : _b.title) || title(item.name.replace(/[-_]/g, " "))
    };
  }).sort((a, b) => {
    if (a.date && b.date) {
      return new Date(b.date).getTime() - new Date(a.date).getTime();
    }
    if (a.date) {
      return -1;
    }
    if (b.date) {
      return 1;
    }
    return a.title.localeCompare(b.title, locale, { numeric: true });
  }).map((item) => [item.name, item.title]);
}

export {
  parseFileName,
  parseJsonFile,
  truthy,
  normalizeMeta,
  sortPages
};
